src/virtual-machine.js

```diff
@@ -378,51 +378,81 @@ class VirtualMachine extends EventEmitter {
         });
     }
 
+    /**
+     * Add a sprite, this could be .sprite2 or .sprite3. Unpack and validate
+     * such a file first.
+     * @param {string | object} input A json string, object, or ArrayBuffer representing the project to load.
+     * @return {!Promise} Promise that resolves after targets are installed.
+     */
+    addSprite (input) {
+        const errorPrefix = 'Sprite Upload Error:';
+        if (typeof input === 'object' && !(input instanceof ArrayBuffer) &&
+          !ArrayBuffer.isView(input)) {
+            // If the input is an object and not any ArrayBuffer
+            // or an ArrayBuffer view (this includes all typed arrays and DataViews)
+            // turn the object into a JSON string, because we suspect
+            // this is a project.json as an object
+            // validate expects a string or buffer as input
+            // TODO not sure if we need to check that it also isn't a data view
+            input = JSON.stringify(input);
+        }
+
+        const validationPromise = new Promise((resolve, reject) => {
+            validate(input, true /* single sprite */, (error, res) => {
+                if (error) {
+                    reject(error);
+                }
+                if (res) {
```

Comment:
If there is no error is there ever an instance where res is false-y?

Reply:
Probably not. =]

Reply:
The thing about removing it is that since reject isn't a return, if there's an error, it executes the reject and then executes the resolve. The actual error still bubbles up, but the fact that the resolve executes after the reject seems weird (and possibly error-prone) to me...